File: Packages/MeterKit/Sources/MeterFeatures/Developer/DeveloperStoreDump.swift
#if DEBUG
import Foundation
import MeterCore

@MainActor
enum DeveloperStoreDump {
    static func json(from dashboard: DashboardModel) -> String {
        let payload = Payload(
            snapshots: dashboard.debugAllReadings().map(SnapshotItem.init),
            subscriptions: dashboard.subscriptions.map(SubscriptionItem.init),
            connections: dashboard.connectionStates().map(ConnectionItem.init)
        )
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        encoder.dateEncodingStrategy = .iso8601
        guard let data = try? encoder.encode(payload),
              let text = String(data: data, encoding: .utf8) else {
            return "{}"
        }
        return text
    }

    private struct Payload: Encodable {
        var snapshots: [SnapshotItem]
        var subscriptions: [SubscriptionItem]
        var connections: [ConnectionItem]
    }

    private struct SnapshotItem: Encodable {
        var accountID: String?
        var providerID: String
        var kind: String
        var fetchedAt: Date
        var periodStart: Date
        var periodEnd: Date
        var currentSpendUSD: String?
        var balanceUSD: String?
        var committedMonthlyUSD: String?
        var chargeDayOfMonth: Int?
        var freeQuotaUsedRatio: Double?
        var dailyDayCount: Int?
        var dailyStart: Date?
        var dailyEnd: Date?

        init(_ snapshot: Snapshot) {
            // 账号才是主键：同一家两份账号的快照只写 providerID 会并成一堆。
            accountID = snapshot.accountID?.rawValue.uuidString
            providerID = snapshot.providerID.rawValue
            kind = snapshot.kind.rawValue
            fetchedAt = snapshot.fetchedAt
            periodStart = snapshot.periodStart
            periodEnd = snapshot.periodEnd
            currentSpendUSD = snapshot.currentSpendUSD.map(Self.decimal)
            balanceUSD = snapshot.balanceUSD.map(Self.decimal)
            committedMonthlyUSD = snapshot.committedMonthlyUSD.map(Self.decimal)
            chargeDayOfMonth = snapshot.chargeDayOfMonth
            freeQuotaUsedRatio = snapshot.freeQuotaUsedRatio
            if let daily = snapshot.dailyUSD, !daily.isEmpty {
                dailyDayCount = daily.count
                dailyStart = daily.keys.min()
                dailyEnd = daily.keys.max()
            }
        }

        private static func decimal(_ money: Money) -> String {
            NSDecimalNumber(decimal: money.usd).stringValue
        }
    }

    private struct SubscriptionItem: Encodable {
        var name: String
        var amountUSD: String
        var period: String
        var accountID: String?
        var providerID: String?

        init(_ subscription: MonthlySubscription) {
            name = subscription.name
            amountUSD = NSDecimalNumber(decimal: subscription.amount.usd).stringValue
            period = subscription.period.rawValue
            accountID = subscription.accountID?.rawValue.uuidString
            providerID = subscription.providerID?.rawValue
        }
    }

    private struct ConnectionItem: Encodable {
        var accountID: String
        var providerID: String
        var nickname: String?
        var isEnabled: Bool
        var credentialReference: String

        init(_ state: ProviderConnectionState) {
            accountID = state.accountID.rawValue.uuidString
            providerID = state.providerID.rawValue
            nickname = state.nickname
            isEnabled = state.isEnabled
            credentialReference = state.credentialReference
        }
    }
}
#endif

File: Android/app/src/main/kotlin/com/zhechengqi/tollcat/developer/DeveloperStoreDump.kt
package com.zhechengqi.tollcat.developer

import com.zhechengqi.tollcat.LedgerStore
import org.json.JSONArray
import org.json.JSONObject

object DeveloperStoreDump {
    fun json(ledger: LedgerStore): String {
        val snapshots = JSONArray()
        for (row in ledger.snapshots()) {
            snapshots.put(
                JSONObject().apply {
                    put("providerID", row.providerId)
                    put("accountID", row.accountId)
                    put("kind", row.kind)
                    put("fetchedAt", row.fetchedAtMillis)
                    put("periodStart", row.periodStartMillis)
                    put("periodEnd", row.periodEndMillis)
                    row.currentSpendUsd?.let { put("currentSpendUSD", it) }
                    row.balanceUsd?.let { put("balanceUSD", it) }
                    row.committedMonthlyUsd?.let { put("committedMonthlyUSD", it) }
                    row.chargeDayOfMonth?.let { put("chargeDayOfMonth", it) }
                    row.freeQuotaUsedRatio?.let { put("freeQuotaUsedRatio", it) }
                },
            )
        }
        val subscriptions = JSONArray()
        for (row in ledger.subscriptions()) {
            subscriptions.put(
                JSONObject().apply {
                    put("name", row.name)
                    put("amountUSD", row.amountUsd)
                    put("period", row.period)
                    row.providerId?.let { put("providerID", it) }
                    row.accountId?.let { put("accountID", it) }
                },
            )
        }
        val connections = JSONArray()
        for (row in ledger.accounts()) {
            connections.put(
                JSONObject().apply {
                    put("providerID", row.providerId)
                    put("accountID", row.accountId)
                    put("credentialReference", row.credentialReference)
                },
            )
        }
        return JSONObject()
            .put("snapshots", snapshots)
            .put("subscriptions", subscriptions)
            .put("connections", connections)
            .toString(2)
    }
}
